File: apps/mobile/flutter/test/unit/application/use_cases/update_todo_use_case_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mocktail/mocktail.dart';
import 'package:habit_todo_app/core/application/use_cases/update_todo_use_case.dart';
import 'package:habit_todo_app/core/domain/entities/todo_entity.dart';
import 'package:habit_todo_app/core/domain/repositories/todo_repository.dart';

class MockTodoRepository extends Mock implements TodoRepository {}

void main() {
  setUpAll(() {
    registerFallbackValue(TodoEntity(
      id: 'fallback',
      title: 'Fallback',
      createdAt: DateTime(2024, 1, 1),
    ));
  });

  group('UpdateTodoUseCase', () {
    late UpdateTodoUseCase useCase;
    late MockTodoRepository mockRepository;

    setUp(() {
      mockRepository = MockTodoRepository();
      useCase = UpdateTodoUseCase(mockRepository);
    });

    test('should update todo with valid input', () async {
      // Arrange
      final existingTodo = TodoEntity(
        id: 'todo_1',
        title: 'Original Title',
        isCompleted: false,
        createdAt: DateTime(2024, 1, 1),
      );

      final updatedTodo = existingTodo.copyWith(
        title: 'Updated Title',
        isCompleted: true,
      );

      when(() => mockRepository.getById('todo_1'))
          .thenAnswer((_) async => existingTodo);
      when(() => mockRepository.update(any()))
          .thenAnswer((_) async => updatedTodo);

      // Act
      final result = await useCase.execute(
        id: 'todo_1',
        title: 'Updated Title',
        isCompleted: true,
      );

      // Assert
      expect(result.title, 'Updated Title');
      expect(result.isCompleted, true);
      verify(() => mockRepository.getById('todo_1')).called(1);
      verify(() => mockRepository.update(any())).called(1);
    });

    test('should update only completion status', () async {
      // Arrange
      final existingTodo = TodoEntity(
        id: 'todo_1',
        title: 'Test Todo',
        isCompleted: false,
        createdAt: DateTime(2024, 1, 1),
      );

      final updatedTodo = existingTodo.copyWith(isCompleted: true);

      when(() => mockRepository.getById('todo_1'))
          .thenAnswer((_) async => existingTodo);
      when(() => mockRepository.update(any()))
          .thenAnswer((_) async => updatedTodo);

      // Act
      final result = await useCase.execute(
        id: 'todo_1',
        isCompleted: true,
      );

      // Assert
      expect(result.isCompleted, true);
      expect(result.title, 'Test Todo'); // Unchanged
      verify(() => mockRepository.update(any())).called(1);
    });

    test('should throw ArgumentError if id is empty', () async {
      // Act & Assert
      expect(
        () => useCase.execute(id: ''),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('should throw ArgumentError if title is empty when provided',
        () async {
      // Act & Assert
      expect(
        () => useCase.execute(id: 'todo_1', title: ''),
        throwsA(isA<ArgumentError>()),
      );
    });

    test('should throw Exception if todo does not exist', () async {
      // Arrange
      when(() => mockRepository.getById('non_existent'))
          .thenAnswer((_) async => null);

      // Act & Assert
      expect(
        () => useCase.execute(id: 'non_existent', title: 'New Title'),
        throwsA(isA<Exception>()),
      );
      verify(() => mockRepository.getById('non_existent')).called(1);
    });

    test('should trim title and description when updating', () async {
      // Arrange
      final existingTodo = TodoEntity(
        id: 'todo_1',
        title: 'Original',
        createdAt: DateTime(2024, 1, 1),
      );

      final updatedTodo = existingTodo.copyWith(
        title: 'Trimmed Title',
        description: 'Trimmed Description',
      );

      when(() => mockRepository.getById('todo_1'))
          .thenAnswer((_) async => existingTodo);
      when(() => mockRepository.update(any(
              that: predicate<TodoEntity>(
            (todo) =>
                todo.title == 'Trimmed Title' &&
                todo.description == 'Trimmed Description',
          )))).thenAnswer((_) async => updatedTodo);

      // Act
      await useCase.execute(
        id: 'todo_1',
        title: '  Trimmed Title  ',
        description: '  Trimmed Description  ',
      );

      // Assert
      verify(() => mockRepository.update(any(
              that: predicate<TodoEntity>(
            (todo) =>
                todo.title == 'Trimmed Title' &&
                todo.description == 'Trimmed Description',
          )))).called(1);
    });

    test('should update deadline', () async {
      // Arrange
      final existingTodo = TodoEntity(
        id: 'todo_1',
        title: 'Test Todo',
        createdAt: DateTime(2024, 1, 1),
      );

      final newDeadline = DateTime(2024, 1, 20, 18, 0);
      final updatedTodo = existingTodo.copyWith(deadline: newDeadline);

      when(() => mockRepository.getById('todo_1'))
          .thenAnswer((_) async => existingTodo);
      when(() => mockRepository.update(any()))
          .thenAnswer((_) async => updatedTodo);

      // Act
      final result = await useCase.execute(
        id: 'todo_1',
        deadline: newDeadline,
      );

      // Assert
      expect(result.deadline, newDeadline);
      verify(() => mockRepository.update(any())).called(1);
    });
  });
}

File: apps/web/test/unit/application/use_cases/update_todo_use_case.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { UpdateTodoUseCase } from "../../../../src/core/application/use_cases/update_todo_use_case";
import type { TodoEntity } from "../../../../src/core/domain/entities/todo_entity";
import type { TodoRepository } from "../../../../src/core/domain/repositories/todo_repository";
import { Priority } from "../../../../src/core/domain/value_objects/priority";

describe("UpdateTodoUseCase", () => {
  let mockRepository: TodoRepository;
  let useCase: UpdateTodoUseCase;

  beforeEach(() => {
    mockRepository = {
      findAll: vi.fn(),
      findById: vi.fn(),
      create: vi.fn(),
      update: vi.fn(),
      delete: vi.fn(),
    };
    useCase = new UpdateTodoUseCase(mockRepository);
  });

  it("should update todo with valid input", async () => {
    const existing: TodoEntity = {
      id: "test-id",
      title: "Original",
      priority: Priority.MEDIUM,
      completed: false,
      createdAt: "2024-01-01T00:00:00.000Z",
      updatedAt: "2024-01-01T00:00:00.000Z",
    };

    const updated: TodoEntity = {
      ...existing,
      title: "Updated",
      updatedAt: "2024-01-02T00:00:00.000Z",
    };

    vi.mocked(mockRepository.findById).mockResolvedValue(existing);
    vi.mocked(mockRepository.update).mockResolvedValue(updated);

    const result = await useCase.execute("test-id", { title: "Updated" });

    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.title).toBe("Updated");
      expect(mockRepository.update).toHaveBeenCalled();
    }
  });

  it("should fail when todo not found", async () => {
    vi.mocked(mockRepository.findById).mockResolvedValue(null);

    const result = await useCase.execute("non-existent", { title: "Updated" });

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Todo not found");
    }
  });

  it("should fail when id is empty", async () => {
    const result = await useCase.execute("", { title: "Updated" });

    expect(result.success).toBe(false);
    if (!result.success) {
      expect(result.error).toBe("Todo ID is required");
    }
  });
});
